Skip computeCompositingRequirements for subtrees that haven't changed and don't contribute overlap.

mattwoodrow opened the pull request:
#### 09fa4e566b17a4001771f00cad52f312b88132f1
<pre>
Skip computeCompositingRequirements for subtrees that haven&apos;t changed and don&apos;t contribute overlap.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281890">https://bugs.webkit.org/show_bug.cgi?id=281890</a>
&lt;<a href="https://rdar.apple.com/138363965">rdar://138363965</a>&gt;

Reviewed by Simon Fraser.

Layers that haven&apos;t changed, don&apos;t have any composited descendant and clip their
descendants shouldn&apos;t need a descendant traversal.

Normally this is done to check for any compositing changes and to accumulate
overlap, but neither of those should be possible with the given conditions

This also removes the anyDescendantHas3DTransform accumulation (which was
depending on visiting all descendants) and replaces it with the existing state
bit computed during update3DTransformedDescendantStatus.

* LayoutTests/compositing/backing/no-backing-for-perspective-expected.txt:
* LayoutTests/platform/ios/compositing/backing/no-backing-for-perspective-expected.txt:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateCompositingLayers):
(WebCore::canSkipComputeCompositingRequirementsForSubtree):
(WebCore::RenderLayerCompositor::computeCompositingRequirements):
(WebCore::RenderLayerCompositor::traverseUnchangedSubtree):
* Source/WebCore/rendering/RenderLayerCompositor.h:

Canonical link: <a href="https://commits.webkit.org/285632@main">https://commits.webkit.org/285632@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5df5ec80ec1e81767ce497601eb8c60eb74d1f38

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73351 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52780 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26159 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/77586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/24589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75466 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60586 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/563 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/77586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/24589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76418 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47655 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/77586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44295 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20591 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/22918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/66148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20944 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79233 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/662 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63121 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16145 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/630 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/660 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/662 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -1165,7 +1164,20 @@ bool RenderLayerCompositor::updateCompositingLayers(CompositingUpdateType update
+    if (layer.needsCompositingRequirementsTraversal() || layer.hasDescendantNeedingCompositingRequirementsTraversal())
+        return false;
+
+    if (!layer.isComposited() || !willBeComposited || layer.hasCompositingDescendant())
```
I'm not 100% convinced the `layer.hasCompositingDescendant()` check is necessary. It seems likely that we could skip over subtrees with compositing layers inside them, I don't think overlap changes from before the current layer can affect descendants (which must not extend outside of this layers overlap bounds due to the clipping).

I also didn't make this a check for stacking context. I think the logic is correct for the paint-order children of this layer. Stacking contexts just have more paint-order children that would be affected by the optimisation.

smfr replied:
I worry about mutations on ancestors or earlier siblings that require us to traverse into the "unchanged" subtree to do things like scrolling tree node registration.

I think it would be safer to constrain this change to stacking contexts for now.

mattwoodrow replied:
If we do have this sort of problem, then how does a stacking context avoid it?

mattwoodrow commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -1165,7 +1164,20 @@ bool RenderLayerCompositor::updateCompositingLayers(CompositingUpdateType update
+    if (!layer.isComposited() || !willBeComposited || layer.hasCompositingDescendant())
+        return false;
+
+    return layer.hasCompositedScrollableOverflow();
```
This could potentially be generalised to some sort of 'the paint-order descendants of this layer don't add new overlap'.

We could accumulate a bool when we do traverse the subtree, checking if the overlap map got mutated usefully, and then use that to know for future passes.

smfr replied:
I'd rather a check for stacking context + non-visible overflow. We want paint containment to trigger this optimization.

mattwoodrow replied:
Non-visible overflow is definitely a good change, will do.

Happy to add stacking context too, but I'm struggling to convince myself that it's needed :)

mattwoodrow commented on LayoutTests/compositing/backing/no-backing-for-perspective-expected.txt:
```diff
@@ -9,23 +9,15 @@ This layer should not have backing store.
-          (position 8.00 8.00)
-          (bounds 342.00 240.00)
+          (position 29.00 47.00)
+          (bounds 342.00 180.00)
```
This test nests perspective inside perspective. The new check for 'has 3d transformed descendants' returns false for the outer, since we've flattened back to 2d at that point (no preserve-3d is present), which I believe is correct.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -1311,49 +1321,50 @@ void RenderLayerCompositor::computeCompositingRequirements(RenderLayer* ancestor
-            didPushOverlapContainer = true;
-            didSpeculativelyPushOverlapContainer = true;
-        }
+    if (!canSkipComputeCompositingRequirementsForSubtree(layer, willBeComposited)) {
```
It's a bit weird to see this here in the flow that isn't `traverseUnchangedSubtree` I'd ask myself which we didn't go down `traverseUnchangedSubtree` instead, which could itself early return.

That suggests that maybe this fix should be about managing the dirty bits, rather than more branching in this logic.

mattwoodrow replied:
The issue is that the `fullPaintOrderTraversalRequired` and  `descendantsRequireCompositingUpdate` compositing state bits can be set.

In the specific case I was debugging, we changed the height of the page, which caused the RenderView's compositing bounds to change and cause `descendantsRequireCompositingUpdate` to be set (via `setSubsequentLayersNeedCompositingRequirementsTraversal()`).
Also, a later sibling layer got removed, which dirties the z-order lists of the shared parent, and sets the `fullPaintOrderTraversalRequired`. That's maybe avoidable with smart tracking of exactly what did change in the z-order lists.

We could use these same conditions to override the checks about whether to go into `traverseUnchangedSubtree`, and then early return inside it?

I think we need to handle the case where something behind the layer has changed, and we really do need to re-evaluate if the layer should be composited or not. If it does still composite, then we can skip over its descendants. That's the intent of the current interception point.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -1165,7 +1164,20 @@ bool RenderLayerCompositor::updateCompositingLayers(CompositingUpdateType update
-void RenderLayerCompositor::computeCompositingRequirements(RenderLayer* ancestorLayer, RenderLayer& layer, LayerOverlapMap& overlapMap, CompositingState& compositingState, BackingSharingState& backingSharingState, bool& descendantHas3DTransform)
+// Unchanged leaf compositing layers that clip their descendants can skip descendant
+// traversal, since their descendants can't contribute any new overlap to the map.
+static bool canSkipComputeCompositingRequirementsForSubtree(RenderLayer& layer, bool willBeComposited)
```
`const RenderLayer& layer`

mattwoodrow committed:
Skip computeCompositingRequirements for subtrees that haven't changed and don't contribute overlap.
https://bugs.webkit.org/show_bug.cgi?id=281890
<rdar://138363965>

Reviewed by Simon Fraser.

Layers that haven't changed, don't have any composited descendant and clip their
descendants shouldn't need a descendant traversal.

Normally this is done to check for any compositing changes and to accumulate
overlap, but neither of those should be possible with the given conditions

This also removes the anyDescendantHas3DTransform accumulation (which was
depending on visiting all descendants) and replaces it with the existing state
bit computed during update3DTransformedDescendantStatus.

* LayoutTests/compositing/backing/no-backing-for-perspective-expected.txt:
* LayoutTests/platform/ios/compositing/backing/no-backing-for-perspective-expected.txt:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateCompositingLayers):
(WebCore::canSkipComputeCompositingRequirementsForSubtree):
(WebCore::RenderLayerCompositor::computeCompositingRequirements):
(WebCore::RenderLayerCompositor::traverseUnchangedSubtree):
* Source/WebCore/rendering/RenderLayerCompositor.h:

Canonical link: https://commits.webkit.org/285632@main
